README.md
Algorithms for (Quadratic) Time-Frequency Distributions
=======================================================

This collection of M-files generate time-frequency distributions
(TFDs) with as few computations and as little memory as possible.
The algorithms are described in [[1]](#references); the discrete TFD definition
they compute is in [[2]](#references).  

There are two sets of algorithms.  The first set (Set1) computes
the exact TFD and minimises on oversampling.  The second set (Set2)
computes a decimated TFD, and therefore is not exact but may be
useful for some applications.

Each set computes four different types of kernels:
- nonseparable kernel
- separable kernel
- Doppler-independent kernel
- lag-independent kernel 
  (see [[1]](#references) for more details)
  
The code is organised as follows:
- folder `full_DTFD` contains the first set of algorithms (computes
  exact TFDs)
- folder `dec_DTFD` contains the second set of algorithms (computes
  decimated TFDs)
- folder `common` and `utils` contain extra functions used by both
  algorithm sets.

These programs should run in either Matlab (v7.9.0) or Octave
(v3.2.4)


## Quick Start
To start, load the paths:

```matlab
>> load_paths_DTFDs;
```

Next, generate a test signal:

```matlab
>> N=128; x=gen_LFM(N,0.1,0.3);
```

Second, generate a TFD with a separable kernel:

```matlab
>> tf=dtfd_sep1(x,{51,'hamm',0,1},{171,'hann'},256,128); 
```

Third, plot the TFD:

```matlab
>> vtfd(tf,x);
```

## Description

1. functions to compute exact TFDs (in folder 'full_DTFD'):
   - `dtfd_DI.m` computes a TFD with Doppler-independent kernel
   - `dtfd_LI.m` computes a TFD with a lag-independent kernel
   - `dtfd_nonsep.m` computes a nonseparable-kernel TFD
   - `dtfd_sep1.m` computes a TFD with a separable kernel
   - `dtfd_sep2.m` computes a TFD with a separable kernel using a
     different algorithm to the `dtfd_sep1.m`

2. functions to compute a decimated TFD (in folder `dec_DTFD`)
   - `dec_dtfd_DI.m` computes a decimated TFD with a
     Doppler-independent kernel
   - `dec_dtfd_LI.m` computes a decimated TFD with a
     lag-independent kernel
   - `dec_dtfd_nonsep.m` computes a decimated TFD with a
     nonseparable kernel
   - `dec_dtfd_sep.m` computes a decimated TFD with a separable kernel
   - `dwvd_grid1.m` computes a decimated Wigner-Ville distribution
   - `dwvd_grid2.m` computes a decimated Wigner-Ville distribution
     with a different decimation grid to that grid in
     `dwvd_grid1.m`

3. Other miscellaneous functions (in folder `utils`) include:
   - `get_analytic.m` computes a discrete analytic signal, as
     defined in [[3]](#references) 
   - `gen_LFM.m` computes a linear frequency modulated (LFM) signal
   - `fft_complex.m` is a function for alternative FFT routines
     (for example to take advantage of parallel computing);
     likewise are `fft_conj_summ.m`, `ifft_complex.m`, and `ifft_conj_symm.m`
        

## Version and Date
  - Version: 0.23
  - Last update: [13-05-2013]


# Examples
In all the examples, start by adding the paths by running the command

 ```matlab
 >> load_paths_DTFDs;
 ```

1. Choi-Williams using the exact TFD algorithm (from Set1 [[1]](#references)):

    ```matlab
    % 1. generate test signal:
    N=256; 
    x=gen_LFM(N,0.05,0.15)+gen_LFM(N,0.2,0.35);
    % 2. generate TFD
    c=dtfd_nonsep(x,'cw',{30}); 
    % 3. plot
    clf; vtfd(c,x);
	  ```

2. Separable-kernel TFD using exact TFD algorithm (from Set1 [[1]](#references))

	  ```matlab
    % 1. generate test signal:
    N=10000; 
    x=gen_LFM(N,0.1,0.3)+gen_LFM(N,0.4,0.1);
    % 2. generate TFD
    Ntime=256; Nfreq=256;
    c=dtfd_sep1(x,{51,'hamm',0,1},{271,'hann'},Ntime,Nfreq); 
    % 3. plot
    clf; vtfd(c,x);
	  ```

3. Doppler-independent kernel TFD using decimated TFD algorithm
   (from Set2 [[1]](#references)) generating selective time portions of the
   signal:

	  ```matlab
    % 1. generate test signal:
    N=171; 
    x=gen_LFM(N,0.1,0.4);
    % 2. generate TFD:
    Nfreq=258; 
    time_dec=[20:3:160,191:250,255,256]; freq_dec=3;
    tf=dec_dtfd_DI(x,{51,'hamm'},Nfreq,time_dec,freq_dec); 
    % 3. plot
    clf; vtfd(tf);
	  ```

4. Separable kernel TFD using decimated TFD algorithm from Set2 [[2]](#references)

    ```matlab
    % 1. generate test signal:
    N=10000; 
    x=gen_LFM(N,0.1,0.3)+gen_LFM(N,0.4,0.1);
    % 2. generate TFD:
    Ntime=512; Nfreq=256;
    time_dec=4; freq_dec=2;
    c=dec_dtfd_sep(x,{51,'hamm',0,1},{271,'hann'}, ...
                   Ntime,Nfreq,time_dec,freq_dec); 
    % 3. plot
    clf; vtfd(c,x);
	  ```


# Test computer setup
- hardware:  Intel Core Duo CPU, 2.13GHz; 2GB memory.
- operating system: Ubuntu GNU/Linux i686 distribution (Natty,
  11.04), with Linux kernel 2.6.38-8-generic
- software: Octave 3.2.4 (using Gnuplot 4.4 patchlevel 2) and
  Matlab (R2009b, R2012a, and R2013a)


# Copyright
```
Copyright (c) 2010 John M. O' Toole, The University of Queensland
All rights reserved.
Email: 	  j.otoole@ieee.org

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following
conditions are met:
    * Redistributions of source code must retain the above
      copyright notice, this list of conditions and the following
      disclaimer.
    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials
      provided with the distribution.
    * Neither the name of the The University of Queensland nor the 
      names of its contributors may be used to endorse or promote 
      products derived from this software without specific prior 
      written permission.

THIS SOFTWARE IS PROVIDED BY JOHN M. O' TOOLE ''AS IS'' AND ANY
EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL JOHN M. O' TOOLE BE
LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
DAMAGE.
``` 

# References
1. J.M. O' Toole and B. Boashash, "Fast and memory-efficient algorithms for computing
   quadratic time–frequency distributions", Applied and Computational Harmonic Analysis,
   vol. 35, no. 2, pp. 350–358, 2013.

2. J.M. Oʼ Toole, M. Mesbah, and B. Boashash, “Improved discrete definition of quadratic
   time–frequency distributions,” IEEE Transactions on Signal Processing, vol. 58,
   Feb. 2010, pp. 906-911.

3. J.M. O' Toole, M. Mesbah, and B. Boashash, "A New Discrete Analytic Signal for Reducing
   Aliasing in the Discrete Wigner-Ville Distribution", IEEE Transactions on Signal
   Processing, vol. 56, no. 11, pp. 5427-5434, Nov. 2008.

4. J.M. Oʼ Toole, M. Mesbah, and B. Boashash, “Algorithms for discrete quadratic
   time–frequency distributions,” WSEAS Transactions on Signal Processing, vol. 4,
   May. 2008, pp. 320-329.


# Contact

John M. O' Toole

Neonatal Brain Research Group,  
Irish Centre for Fetal and Neonatal Translational Research (INFANT),  
Department of Paediatrics and Child Health,  
University College Cork,  
Western Gateway Building, Room 2.17,  
Cork, Ireland

- email: j.otoole -AT- ieee.org
- web: http://otoolej.github.io/


ChangeLog.org
#+TITLE: Change log for fast-TFD algorithms
#
# started: 23-Jan-2013

* [2016-08-11 Thu] 
  + replaced README.txt with README.md

* [2013-05-13 Mon]
  + changed 'getWin' to 'get_window' function as was conflicting with TFSA package

* [2013-05-08 Wed]
  + minor fixes to window functions in kernel

* [2013-01-23 Wed]
  + added the argument 'n' to the lag window function to normalise lag window (g2), e.g.:

    >> dtfd_sep1(x,{floor(256/8),'rect',0,1},{11,'hann',0,0,'n'},256,128);

* [2013-03-29 Fri]
  + can use 'time-reversed' lag window, e.g.

    >> dtfd_sep1(x,{floor(256/8),'rect',0,1},{128,'bartlett,0,0,'n',1},256,128);


dec_DTFD/dec_dtfd_nonsep.m
%-------------------------------------------------------------------------------
% Decimated TFD (time-frequency distribution)
% Kernel Type:  nonseparable
%
%
% USE: [tf,g] = dec_dtfd_nonsep(z,tfd_type,tfd_params,time_dec,freq_dec)
%
% INPUT: 
%      z = input signal (either real-valued signal of length-N or
%          complex-valued analytic signal of length-2N)
%
%      tfd_type = { 'wvd' | 'swvd' | 'pwvd' | 'sep' | 'cw' | 'mb' }
%            wvd  - Wigner-Ville distribution
%            swvd - Smoothed Wigner-Ville distribution
%                   (lag-independent kernel)
%            pwvd - Pseudo Wigner-Ville distribution
%                   (Doppler-independent kernel)
%            sep  - Separable-kernel distribution 
%                   (combintation of SWVD and PWVD)
%            cw   - Choi-Williams distribution
%            mb   - Modified-B distribution
% 
%      tfd_params = cell of kernel parameters:
%            wvd  - {}
%            swvd - {win_length,win_type,[win_param]}
%                   e.g. {11,'hamm'}
%            pwvd - {win_length,win_type,[win_param]}
%                   e.g. {200,'cosh',0.1
%            sep  - { {win1_length,win1_type,[win1_param]}, 
%                    {win2_length,win2_type,[win2_param]}
%                   where win1 is the doppler window and win2 is the 
%                   lag window, e.g. { {11,'hamm'}, {200,'cosh',0.1} }
%            cw   - {sigma_parameter}
%            mb   - {beta_parameter} in the range 1<beta<0
%
%      time_dec = integer value time decimation value
%                 where (2N/time_dec) must be an integer
%
%      freq_dec = integer value frequency decimation value
%                 where (N/freq_dec) must be an integer
%      
% OUTPUT: 
%      tf = LxJ TFD, where L=(2N/time_dec) and J=(N/freq_dec)
%      g  = kernel
%
% EXAMPLE:
%      N=10000; 
%      time_dec=160; freq_dec=80;
%      x=gen_LFM(N,0.1,0.3)+gen_LFM(N,0.4,0.1);
%
%      c=dec_dtfd_nonsep(x,'cw',{100},time_dec,freq_dec); 
%      vtfd(c,x);


%
%  Copyright (c) 2010, John M. O' Toole, The University of Queensland
%  All rights reserved.
%
%  Redistribution and use in source and binary forms, with or without
%  modification, are permitted provided that the following
%  conditions are met:
%      * Redistributions of source code must retain the above
%        copyright notice, this list of conditions and the following
%        disclaimer.
%      * Redistributions in binary form must reproduce the above
%        copyright notice, this list of conditions and the following
%        disclaimer in the documentation and/or other materials
%        provided with the distribution.
%      * Neither the name of the The University of Queensland nor the 
%        names of its contributors may be used to endorse or promote 
%        products derived from this software without specific prior 
%        written permission.
%  
%  THIS SOFTWARE IS PROVIDED BY JOHN M. O' TOOLE ''AS IS'' AND ANY
%  EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
%  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
%  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL JOHN M. O' TOOLE BE
%  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
%  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
%  OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
%  BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
%  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
%  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
%  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
%  DAMAGE.
%-------------------------------------------------------------------------------
function [tf,g] = dec_dtfd_nonsep(z,tfd_type,tfd_params,time_dec,freq_dec)
if(nargin<4) time_dec=1; end
if(nargin<5) freq_dec=1; end
[z,N2,N,Nh]=check_analytic_signal(z);
tf=[]; R=[]; g=[];


global_vars;
set_DBvars('all',0); set_DBvars('dbtfd',0);


%-------------------------------------------------------------------------------
% 1. Check decimation parameters
%-------------------------------------------------------------------------------
[L,time_dec]=checkParams_seq(time_dec,N2);
[J,freq_dec]=checkParams_seq(freq_dec,N);
Lh=ceil(L/2); Jh=ceil(J/2); J2=2*J;
if(FAST_DTFD_DBvars) dispVars(N2,N,time_dec,freq_dec,L,Lh,J,J2,Jh); end


%-------------------------------------------------------------------------
% 3. Get TL kernel Kdown[n,m]
%-------------------------------------------------------------------------
nt=0:time_dec:N2-1;
ni_even=find(rem(nt,2)==0)-1;
ni_odd=find(rem(nt,2)==1)-1;
Le=length(ni_even); Lo=length(ni_odd);

if(rem(time_dec,2)==0)
  time_dec_h=time_dec/2;  
else
  time_dec_h=time_dec;  
end
if(FAST_DTFD_DBvars) dispVars(Le,Lo,time_dec_h); end


Racross=zeros(Le+Lo,J);
AFf_EvenM=zeros(Le,1);
AFf_OddM=zeros(Lo,1);


%-------------------------------------------------------------------------
% For m=even
%-------------------------------------------------------------------------
n=0:N-1; m1=1:Jh-1;
for m=0:Jh
  AFslice_EvenM=zeros(N,1);  
  
  %------------------------------------
  % Fold in the lag direction
  %------------------------------------
  for p=0:freq_dec-1
      g=gen_kern_lag_slice( 2*(p*J+m),tfd_type,tfd_params,N);
      AFslice_EvenM = AFslice_EvenM + ...
          fft(getKslice_EvenM(z,n,p.*J+m,N2,N,Nh)) .* g;
  end

  %------------------------------------
  % Fold in the Doppler direction
  %------------------------------------
  AFf_EvenM=fold_sig_full(AFslice_EvenM,Le,time_dec_h);
  
  
  %------------------------------------
  % DFT the lag slice to the time--lag 
  % domain.
  %------------------------------------
  Racross(ni_even+1,m+1)=ifft_complex(AFf_EvenM);

end


%-------------------------------------------------------------------------
% For m=odd  (only if time_dec is odd value, otherwise no m=odd!)
%-------------------------------------------------------------------------
if(Lo>0)
  m2=0:Jh-1;
  if(FAST_DTFD_DBvars) dispVars((time_dec_h-1)/2); end
  for m=0:Jh
    AFslice_OddM=zeros(N,1);  
    
    %------------------------------------
    % Fold in the lag direction
    %------------------------------------
    for p=0:freq_dec-1
        g=gen_kern_lag_slice(2*(p*J+m)+1,tfd_type,tfd_params,N);        
        AFslice_OddM = AFslice_OddM + ...
            fft(getKslice_OddM(z,n,p.*J+m,N2,N,Nh)) .* g;
    end

    %------------------------------------
    % Fold in the Doppler direction
    %------------------------------------
    if(time_dec>1)
      AFf_OddM=fold_sig_full_odd(AFslice_OddM,Lo,N,time_dec_h,...
                                 (time_dec_h-1)/2);  
    else
      AFf_OddM=AFslice_OddM;
    end
    

    %------------------------------------
    % DFT the lag slice to the time--lag 
    % domain.
    %------------------------------------
    Racross(ni_odd+1,m+1)=ifft_complex(AFf_OddM);
    
  end
end


%-------------------------------------------------------------------------
% Then recover negative lag values from positive ones
%-------------------------------------------------------------------------
m1=1:Jh-1; m2=0:Jh-1;
Racross(ni_even+1,J-m1+1)=conj( Racross(ni_even+1,m1+1) );
if(~isempty(ni_odd))
  Racross(ni_odd+1,J-m2)=conj( Racross(ni_odd+1,m2+1) );
end



tf=dft_Racross(Racross,L,J,N,ni_even,ni_odd);
tf=tf./N2;  


% final error check and return real TFD:
tf=istfd_real(tf,z,'dec_dtfd_nonsep');











function Kslice=getKslice(z,m,N,N2)
%-------------------------------------------------------------------------------
% Get TL function for one value of m, from n=0,...,N-1.
% (Kslice is Kdown[n,m_0])
%-------------------------------------------------------------------------------

Kslice=zeros(N,1);
n=0:N-1;
if(rem(m,2)==0)
  m=m/2;
  i1=mod(n+m,N2); i2=mod(n-m,N2);
  Kslice=z(i1+1).*conj(z(i2+1));
else
  m=(m-1)/2;
  i3=mod(n+m+1,N2); i2=mod(n-m,N2);
  Kslice = z(i3+1).*conj(z(i2+1)); 
end


function Kslice=getKslice_EvenM(z,n,m,N2,N,Nh)
%-------------------------------------------------------------------------------
% Get TL function for one value of m, from n=0,...,N-1.
% (Kslice is Kdown[n,m_0])
%-------------------------------------------------------------------------------
if(m<=Nh)
  i1=mod(n+m,N2); i2=mod(n-m,N2);
  Kslice=z(i1+1).*conj(z(i2+1));
else
  m=N-m;
  i1=mod(n+m,N2); i2=mod(n-m,N2);
  Kslice=conj(z(i1+1)).*z(i2+1);
end

  

function Kslice=getKslice_OddM(z,n,m,N2,N,Nh)
%-------------------------------------------------------------------------------
% Get TL function for one value of m, from n=0,...,N-1.
% (Kslice is Kdown[n,m_0])
%-------------------------------------------------------------------------------
if(m<Nh)
  i1=mod(n+m+1,N2); i2=mod(n-m,N2);
  Kslice=z(i1+1).*conj(z(i2+1));
else
  m=N-m-1;
  i1=mod(n+m+1,N2); i2=mod(n-m,N2);
  Kslice=conj(z(i1+1)).*z(i2+1);
end



